### PR TITLE
[cli] Add flags to start gcs only or dashboard only.

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

This PR add the ability to the ray cli to start specific component.

```
ray start --gcs --raylet --api-server
```

`--gcs` will start the gcs server
`--raylet` will start the raylet
`--api-server` will start the dashboard

This flags can be used together or separately, but `--api-server` has to be used when `--raylet` is used for now.

Non of these flags can be used with `--head`.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_ray_deploy.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    },
+)
+
+
+@pytest.fixture
+def docker_cluster(gcs_node, rayelt_1, rayelt_2):
+    yield (gcs_node, rayelt_1, rayelt_2)
+
+
+def test_ray_deploy_e2e(docker_cluster, tmp_path):
```
Comment (iycheng) [inline python/ray/tests/test_ray_deploy.py:67]: @allenyin55  @yiranwang52  please check this script to see what and how.

**Review by yiranwang52 — APPROVED** [review]

**Review by allenyin55 — APPROVED** [review]

**Comment by scv119** [comment]
currently we have following options for `ray start`
1. `ray start` start worker node, including raylet and agent
2. `ray start --head` start header node, including gcs, raylet, agent and dashboard
3. `ray start --head --include-dashboard` start api server on **head node only**; this has to be 

in this case, does it make sense make following changes?

1. `ray start --gcs-only` only start gcs.
2. `ray start --include-dashboard` now allows to be call on worker node as well

**Comment by iycheng** [comment]
It seems the client setup got messed up. I'm looking.

**Comment by jjyao** [comment]
I'm most curious how this should work with existing flags, especially with `--head`:
* Should `--gcs` always come with `--head` indicating that it starts on the head node?
* Should `--api-server` always come with `--head`?
* Should `--raylet` always come with `--head`, is it sensible to use this option to start raylet on the worker node?
* What about other processes like agent, log monitor, are they started with `--raylet`?

**Comment by iycheng** [comment]
@scv119 
> ray start --include-dashboard now allows to be call on worker node as well

include-dashboard means the UI, not means the api server. I think we start the api server anyway. So what's why I use `--api-server`.

**Comment by iycheng** [comment]
@jjyao it's not allowed. `--head` starts several things. `--gcs/--raylet/...` starts the component by themselves. No need to allow the combination.

**Comment by iycheng** [comment]
@jjyao  log-monitor should always come with raylet, otherwise, log streaming will fail. (we always start log-monitor with raylet right now, no change in the behavior).

**Comment by iycheng** [comment]
@jjyao in summary, nothing changed with existing flags their semantics are the same as before.

**Comment by jjyao** [comment]
What's the difference between using `ray start` vs `ray start --raylet` to start a worker node?

**Comment by iycheng** [comment]
@jjyao  no difference.

**Review by ericl — CHANGES_REQUESTED** [review]
I don't think this is an API we should expose to developers. Ray internals should not be something infra engineers should be aware of. For this change to proceed, I'd have to see proof that there is no other way to achieve the goals besides this.

As I understand, the goal is prioritization of the GCS process over other internal Ray processes. There are other ways of achieving this that don't require exposing a public API.

**Comment by rkooo567** [comment]
Btw, I want to make sure "log_monitor" and "dashboard_agent" is deployed together when they are started standalone. Without that we will lose error logs from GCS and all metrics emitted to dashboard + gcs

**Comment by SongGuyang** [comment]
In Ant, our requirement is no raylet in head node.  The worker processes which is brought by raylet will affect the stability of GCS and dashboard, e.g. cpu, memory, disk, network and so on.

**Comment by ericl** [comment]
@SongGuyang , the goal makes sense. How about adding a --disable-scheduling flag for the head node? The implementation detail can either be not starting raylet or starting it with scheduling disabled.

**Comment by SongGuyang** [comment]
So, can we support a way as follows:
- start a head node in which doesn't contain raylet by default: `ray start --head`
- start a worker node: `ray start`
- start a standalone cluster which contains all the core processes(raylet, gcs, dashboard, agent...): `ray start --standalone`

**Comment by SongGuyang** [comment]
> --disable-scheduling

This flag seems also good for us.

**Comment by ericl** [comment]
The reason to have a new flag is backwards compatibility... we'd break all existing ray deployments otherwise.

**Comment by SongGuyang** [comment]
I see. Keeping backwards compatibility makes sense.

**Pull request closed without merging**